### PR TITLE
Support for lchown and fchownat

### PR DESCRIPTION
**Pull request by vtikoo**
### Summary
- Support for fchownat and lchown syscalls.
- symlink and lstat hostfs operations now propagate identities.
- Path to inode lookup in ramfs and ext2fs was returning ENOENT if an intermediate component is not a directory. Changed it to return ENOTDIR.

### CODE REVIEW

File: ext2/ext2.c [hunk 1]
```diff
@@ -5324,6 +5329,62 @@ static int _ext2_fchown(
...
+        &tfs));
+    if (tfs)
+    {
+        /* delegate operation to target filesystem */
+        ECHECK((ret = tfs->fs_lchown(tfs, locals->suffix, owner, group)));
+        goto done;
+    }
+
+    if (!S_ISLNK(locals->inode.i_mode))
+        ERAISE(-ENOTDIR);
```
Comment (jxyang) [inline ext2/ext2.c:5373]: Does this mean we return -ENOTDIR when the file is not a soft link? I think in that case, we just treat it as a regular file.

File: kernel/ramfs.c [hunk 2]
```diff
@@ -534,8 +534,10 @@ static int _path_to_inode_recursive(
     {
         for (size_t i = 0; i < ntoks; i++)
         {
-            inode_t* p;
+            if (!S_ISDIR(parent->mode))
```
Comment (jxyang) [inline kernel/ramfs.c:537]: When would this happen?

File: tests/ltp/Makefile [hunk 3]
```diff
@@ -118,22 +119,27 @@ group-file:
 	echo "grp701:x:701:user700" >> appdir/etc/group
 	echo "grp704:x:704:" >> appdir/etc/group
 	echo "grp705:x:705:" >> appdir/etc/group
+	echo "grp1000:x:1000:" >> appdir/etc/group
 
 add_host_users_and_groups:
 	sudo groupadd -g 100701 grp701
 	sudo groupadd -g 100704 grp704
 	sudo groupadd -g 100705 grp705
+	sudo groupadd -g 101000 grp1000
```
Comment (jxyang) [inline tests/ltp/Makefile:128]: Sometimes the group might already be there from previous build. We should either append ` | ture` or check whether the group/user exists before the add commands.

File: kernel/uid_gid.c [hunk 4]
```diff
@@ -880,6 +881,26 @@ int check_thread_group_membership(gid_t group)
...
 }
 
+static int _path_checks(const char* pathname)
+{
+    int ret = 0;
+
+    if (!pathname)
+        return -EINVAL;
+
+    if (pathname == (void*)0xffffffffffffffff)
```
Comment (jxyang) [inline kernel/uid_gid.c:891]: Use we use `-1`?

File: kernel/uid_gid.c [hunk 5]
```diff
@@ -1000,3 +1020,50 @@ long myst_syscall_fchown(int fd, uid_t owner, gid_t group)
 
     return ret;
 }
+
+long myst_syscall_lchown(const char* pathname, uid_t owner, gid_t group)
```
Comment (jxyang) [inline kernel/uid_gid.c:1024]: This, `myst_syscall_fchown`, `myst_syscall_chown`, and `myst_syscall_fchownat` should be put together in a new file `chown.c`.

File: kernel/syscall.c [hunk 6]
```diff
@@ -1881,6 +1881,95 @@ long myst_syscall_fstatfs(int fd, struct statfs* buf)
...
+        ERAISE(-EINVAL);
+
+    if ((flags & ~(AT_EMPTY_PATH | AT_SYMLINK_NOFOLLOW)) != 0)
+        ERAISE(-EINVAL);
+
+    if (!(locals = malloc(sizeof(struct locals))))
+        ERAISE(-ENOMEM);
+
+    /* If pathname is absolute, then ignore dirfd */
+    if (*pathname == '/' || dirfd == AT_FDCWD)
```
Comment (jxyang) [inline kernel/syscall.c:1909]: I think all these cases can be consolidated with:
```c
   helper_function_to_get_real_path(dirfd, pathname, &locals->realpath);
   ret = (flags & AT_SYMLINK_NOFOLLOW) ? 
                myst_syscall_lchown(locals->realpath, owner, group) : 
                myst_syscall_chown(locals->realpath, owner, group);
```

File: kernel/ramfs.c (hunk 2 above)
Comment (vtikoo) [reply kernel/ramfs.c:537]: If any intermediate path component is a file. For eg: /reg-file/file2. On linux this returns on ENOTDIR, mystikos was returning an ENOENT.

File: kernel/uid_gid.c (hunk 4 above)
Comment (vtikoo) [reply kernel/uid_gid.c:891]: This really should be a bad address check. One of the lchown ltp tests uses `0xffffffffffffffff` as a bad address test.

File: ext2/ext2.c (hunk 1 above)
Comment (vtikoo) [reply ext2/ext2.c]: Thanks for pointing this out. Fixed.

File: tests/ltp/Makefile (hunk 3 above)
Comment (vtikoo) [reply tests/ltp/Makefile]: Fixed.

File: kernel/syscall.c (hunk 6 above)
Comment (vtikoo) [reply kernel/syscall.c]: Thanks for the suggestion, have incorporated it.

File: kernel/uid_gid.c (hunk 5 above)
Comment (vtikoo) [reply kernel/uid_gid.c]: Done.

File: include/myst/syscall.h [hunk 7]
```diff
@@ -317,5 +317,18 @@ long myst_syscall_getcpu(unsigned* cpu, unsigned* node);
...
 long myst_syscall_lchown(const char* pathname, uid_t owner, gid_t group);
+long myst_syscall_fchownat(
+    int dirfd,
+    const char* pathname,
+    uid_t owner,
+    gid_t group,
+    int flags);
+
+/* Used by XXXXXat() syscalls */
+long get_absolute_path_from_dirfd(
```
Comment (jxyang) [inline include/myst/syscall.h:328]: I think the convention is to prefix non-static function names with "myst_".

Comment (vtikoo) [reply include/myst/syscall.h]: Renamed.

**Review by jxyang — APPROVED** [review]

**Pull request merged**